fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/service/TwoFactorServiceImpl.java

```diff
@@ -0,0 +1,221 @@
+        this.smsMessageScheduledJobService = smsMessageScheduledJobService;
+        this.otpRequestRepository = otpRequestRepository;
+        this.tfAccessTokenRepository = tfAccessTokenRepository;
+        this.smsMessageRepository = smsMessageRepository;
+        this.configurationService = configurationService;
+    }
+
+
+    @Override
+    public List<OTPDeliveryMethod> getDeliveryMethodsForUser(final AppUser user) {
+        List<OTPDeliveryMethod> deliveryMethods = new ArrayList<>();
+
+        OTPDeliveryMethod smsMethod = getSMSDeliveryMethodForUser(user);
+        if(smsMethod != null) {
+            deliveryMethods.add(smsMethod);
+        }
+        OTPDeliveryMethod emailDelivery = getEmailDeliveryMethodForUser(user);
+        if(emailDelivery != null) {
+            deliveryMethods.add(emailDelivery);
+        }
+
+        return deliveryMethods;
+    }
+
+    @Override
+    public OTPRequest createNewOTPToken(final AppUser user, final String deliveryMethodName,
+                                        final boolean extendedAccessToken) {
+        if(TwoFactorConstants.SMS_DELIVERY_METHOD_NAME.equalsIgnoreCase(deliveryMethodName)) {
+            OTPDeliveryMethod smsDelivery = getSMSDeliveryMethodForUser(user);
+            if(smsDelivery == null) {
+                throw new OTPDeliveryMethodInvalidException();
+            }
+            final OTPRequest request = generateNewToken(smsDelivery, extendedAccessToken);
+            final String smsText = configurationService.getFormattedSmsTextFor(user, request);
+            SmsMessage smsMessage = SmsMessage.pendingSms(null, null, null, user.getStaff(), smsText,
+                    user.getStaff().mobileNo(), null);
+            this.smsMessageRepository.save(smsMessage);
+            smsMessageScheduledJobService.sendTriggeredMessage(Collections.singleton(smsMessage),
+                    configurationService.getSMSProviderId());
+            otpRequestRepository.addOTPRequest(user, request);
+            return request;
+        } else if(TwoFactorConstants.EMAIL_DELIVERY_METHOD_NAME.equalsIgnoreCase(deliveryMethodName)) {
+            OTPDeliveryMethod emailDelivery = getEmailDeliveryMethodForUser(user);
+            if(emailDelivery == null) {
+                throw new OTPDeliveryMethodInvalidException();
+            }
+            final OTPRequest request = generateNewToken(emailDelivery, extendedAccessToken);
+            final String emailSubject = configurationService.getFormattedEmailSubjectFor(user, request);
+            final String emailBody = configurationService.getFormattedEmailBodyFor(user, request);
+            final EmailDetail emailData = new EmailDetail(emailSubject, emailBody, user.getEmail(),
+                    user.getFirstname() + " " + user.getLastname());
+            emailService.sendDefinedEmail(emailData);
+            otpRequestRepository.addOTPRequest(user, request);
+            return request;
+        } else {
```

Comment:
Without the else, no error would be returned if a request for OTP is sent for undefined delivery method. Would you like me to do the validation somewhere else?